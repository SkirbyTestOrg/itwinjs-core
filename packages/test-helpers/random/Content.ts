/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2017 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import * as faker from "faker";
import * as c from "../../common/src/content";
import { createRandomECClassInfo, createRandomRelationshipPath } from "./EC";
import { nullable } from "./Misc";

const createRandomSelectClassInfo = (): c.SelectClassInfo => {
  return {
    selectClassInfo: createRandomECClassInfo(),
    isSelectPolymorphic: faker.random.boolean(),
    pathToPrimaryClass: createRandomRelationshipPath(),
    relatedPropertyPaths: [createRandomRelationshipPath(1), createRandomRelationshipPath(1)],
  } as c.SelectClassInfo;
};

export const createRandomCategory = (): c.CategoryDescription => {
  return {
    name: faker.random.word(),
    label: faker.random.words(),
    description: faker.lorem.sentence(),
    priority: faker.random.number(),
    expand: faker.random.boolean(),
  } as c.CategoryDescription;
};

export const createRandomPrimitiveTypeDescription = (): c.TypeDescription => {
  return {
    valueFormat: c.PropertyValueFormat.Primitive,
    typeName: faker.database.type(),
  } as c.PrimitiveTypeDescription;
};

export const createRandomEditorDescription = (): c.EditorDescription => {
  return {
    name: faker.random.word(),
  } as c.EditorDescription;
};

const createRandomPrimitiveFieldJson = () => {
  return {
    category: createRandomCategory(),
    name: faker.random.word(),
    label: faker.random.words(),
    type: createRandomPrimitiveTypeDescription(),
    isReadonly: faker.random.boolean(),
    priority: faker.random.number(),
    editor: nullable(createRandomEditorDescription),
  };
};

export const createRandomPrimitiveField = (): c.Field => {
  return c.Field.fromJSON(createRandomPrimitiveFieldJson())!;
};

export const createRandomDescriptorJson = (displayType?: string) => {
  const selectClasses = [createRandomSelectClassInfo(), createRandomSelectClassInfo()];
  const fields = [createRandomPrimitiveFieldJson(), createRandomPrimitiveFieldJson(), createRandomPrimitiveFieldJson()];
  return {
    connectionId: faker.random.uuid(),
    inputKeysHash: faker.random.uuid(),
    contentOptions: faker.random.objectElement(),
    displayType: displayType || faker.lorem.words(),
    selectClasses,
    fields,
    contentFlags: 0,
  };
};

export const createRandomDescriptor = (displayType?: string): c.Descriptor => {
  return c.Descriptor.fromJSON(createRandomDescriptorJson(displayType))!;
};
