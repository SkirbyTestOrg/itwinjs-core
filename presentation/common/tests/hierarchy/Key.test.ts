/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import * as faker from "faker";
import {
  fromJSON, StandardNodeTypes, BaseNodeKey,
  ECInstanceNodeKeyJSON, ECClassGroupingNodeKey,
  ECPropertyGroupingNodeKey, LabelGroupingNodeKey,
} from "../../lib/hierarchy/Key";
import { createRandomId } from "../_helpers/random";

describe("NodeKey fromJSON", () => {

  it("creates BaseNodeKey", () => {
    const json: BaseNodeKey = {
      type: faker.random.word(),
      pathFromRoot: [faker.random.uuid()],
    };
    const key = fromJSON(json);
    expect(key).to.matchSnapshot();
  });

  it("creates ECInstanceNodeKey", () => {
    const json: ECInstanceNodeKeyJSON = {
      type: StandardNodeTypes.ECInstanceNode,
      pathFromRoot: [faker.random.uuid()],
      instanceKey: {
        className: faker.random.word(),
        id: createRandomId().toString(),
      },
    };
    const key = fromJSON(json);
    expect(key).to.matchSnapshot();
  });

  it("creates ECClassGroupingNodeKey", () => {
    const json: ECClassGroupingNodeKey = {
      type: StandardNodeTypes.ECClassGroupingNode,
      pathFromRoot: [faker.random.uuid()],
      className: faker.random.word(),
    };
    const key = fromJSON(json);
    expect(key).to.matchSnapshot();
  });

  it("creates ECPropertyGroupingNodeKey", () => {
    const json: ECPropertyGroupingNodeKey = {
      type: StandardNodeTypes.ECPropertyGroupingNode,
      pathFromRoot: [faker.random.uuid()],
      className: faker.random.word(),
      propertyName: faker.random.word(),
      groupingValue: faker.random.number(),
    };
    const key = fromJSON(json);
    expect(key).to.matchSnapshot();
  });

  it("creates LabelGroupingNodeKey", () => {
    const json: LabelGroupingNodeKey = {
      type: StandardNodeTypes.DisplayLabelGroupingNode,
      pathFromRoot: [faker.random.uuid()],
      label: faker.random.words(),
    };
    const key = fromJSON(json);
    expect(key).to.matchSnapshot();
  });

});
