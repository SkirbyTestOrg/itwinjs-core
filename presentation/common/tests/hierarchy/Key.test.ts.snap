// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeKey fromJSON creates BaseNodeKey 1`] = `
Object {
  "pathFromRoot": Array [
    "d75d7bdc-c0f7-4e4a-9b67-6575afae7eb8",
  ],
  "type": "backing up",
}
`;

exports[`NodeKey fromJSON creates ECClassGroupingNodeKey 1`] = `
Object {
  "className": "New Mexico",
  "pathFromRoot": Array [
    "f6197eae-8ae8-44c3-b286-3432422b5f98",
  ],
  "type": "ECClassGroupingNode",
}
`;

exports[`NodeKey fromJSON creates ECInstanceNodeKey 1`] = `
Object {
  "instanceKey": Object {
    "className": "Ameliorated",
    "id": "0x451a00000068da",
  },
  "pathFromRoot": Array [
    "000df554-2edd-4edd-8b22-be522c878f31",
  ],
  "type": "ECInstanceNode",
}
`;

exports[`NodeKey fromJSON creates ECPropertyGroupingNodeKey 1`] = `
Object {
  "className": "Division",
  "groupingValue": 77874,
  "pathFromRoot": Array [
    "e6a23e27-348a-47e3-9061-62dfaca34d39",
  ],
  "propertyName": "Planner",
  "type": "ECPropertyGroupingNode",
}
`;

exports[`NodeKey fromJSON creates LabelGroupingNodeKey 1`] = `
Object {
  "label": "evolve Berkshire",
  "pathFromRoot": Array [
    "f1dd9cea-1a68-494b-a67f-635d9d40589b",
  ],
  "type": "DisplayLabelGroupingNode",
}
`;
